### PR TITLE
Added some useful methods to Cake\Network\Request

### PR DESCRIPTION
**Pull request by mirko-pagliai**
These are some useful methods that I use in my applications.
Why?

Suppose we want to execute a code for all actions except those of the `Stuff` controller.

`AppController.php`:

```
public function beforeFilter(\Cake\Event\Event $event) {
    if(!$this->request->isController('Stuff')) {
        // do something...
    }
}
```

Or I want to run a code to all the "delete" actions:

```
public function beforeFilter(\Cake\Event\Event $event) {
    if($this->request->isAction('delete')) {
        // do something...
    }
}
```

And so on...

### CODE REVIEW

**Comment by markstory** [comment]
Why not use the Request methods that allow `is()` to be extended?

```php
// in bootstrap
Request::addDetector('controller', function ($val, $request) {
  return $val == $request->param('controller');
});

$request->isController('Articles');
```

I would prefer to not add additional methods that can already be added through existing extension points.

**Comment by mirko-pagliai** [comment]
Yes you are right. I will modify my PR.

**Comment by mirko-pagliai** [comment]
Yes @markstory you are right. Where do you suggest to add the code? `Bootstrap` of app?

**Comment by thinkingmedia** [comment]
I don't understand why that would be better then

```
public function beforeFilter(\Cake\Event\Event $event) {
    if($event->subject instanceof StuffController) {
        // do something...
    }
}
```

**Comment by mirko-pagliai** [comment]
And for other methods?

**Comment by thinkingmedia** [comment]
@mirko-pagliai all the other methods are very handy :+1:

**Comment by mirko-pagliai** [comment]
@markstory from [docs](http://book.cakephp.org/3.0/en/controllers/request-response.html#checking-request-conditions):

> The callback will receive the request object as its only parameter.

In fact, the code that you suggested will not work.

File: src/Network/Request.php [hunk 1]
```diff
@@ -767,6 +768,73 @@ public function isAll(array $types)
...
+     * Check if the specified prefix matches the current prefix.
+     *
+     * @param string|array $prefix Prefix to check.
+     * @return bool Success.
+     */
+    public function isPrefix($prefix)
+    {
+        if (is_array($prefix)) {
+            return in_array($this->param('prefix'), $prefix);
+        } else {
```
Comment (ADmad) [inline src/Network/Request.php]: The `else` is not needed since there's a `return` in `if` block.

Comment (mirko-pagliai) [inline src/Network/Request.php]: You're right.

Comment (thinkingmedia) [inline src/Network/Request.php]: I would just use `?:` and a single return.

Comment (mirko-pagliai) [inline src/Network/Request.php]: Yes. But I think as well the code is more readable and understandable.

**Comment by markstory** [comment]
@mirko-pagliai Good point, for some reason I thought detectors supported arguments. Given that all of these methods are for interacting with params, would it useful to generalize the code? Something like:

```php
$request->hasParam('controller', 'Articles');
$request->hasParam('action', 'view');
$request->hasParam('prefix', 'admin');
```

**Comment by mirko-pagliai** [comment]
@markstory Yes, good idea. But I would choose another name. One method called `hasParam()` makes me think about a method that checks whether the request has that parameter.

In other words, `hasParam()` makes me think something like `(bool) $request->param($name)` (yes, this param is not empty!).

But this now gives me an idea ...

What about this?

    public method hasParam($name, $value = null) {
        if (empty($value) {
            return (bool) $this->param($name);
        }

        if (is_array($value) {
            return in_array($this->param($name), $value);
        }

        return $this->param($name) === $value;
    }

So, whit this:

    if ($this->request->has('prefix')) {
        echo "Yes, this request as a prefix! I don't know which, I just want to know if it has any prefix";
    }

    if ($this->request->has('prefix', 'admin') {
        echo "Yes, this request as a prefix! The prefix is admin";
    }

    if ($this->request->has('prefix', ['admin', 'manager']) {
        echo "Yes, this request as a prefix! The prefix is admin or manager";
    }

Other idea: instead of a boolean, return the parameter value founded if this is correct, otherwise `FALSE`.

**Comment by markstory** [comment]
I would be cautious about calling the method `has()` as there could be `hasHeader()` and `hasServer()` so keeping it constrained to routing parameters somehow would be nice.

**Comment by dakota** [comment]
`isParam()` perhaps?

**Comment by mirko-pagliai** [comment]
Sorry, it was `hasParam()`, see the code :-) (the examples are wrong)

**Comment by jadb** [comment]
Why not consider adding support for arguments in detectors? Method signature will change by adding a new optional param, but that's ok, no? I personally find the end result will be more inline with the way other things are done as @markstory previously suggested.

**Comment by jadb** [comment]
Something like this (ran existing tests but didn't write new ones yet):

```diff
diff --git i/src/Network/Request.php w/src/Network/Request.php
index 5b3793d..111fdf5 100644
--- i/src/Network/Request.php
+++ w/src/Network/Request.php
@@ -615,7 +615,7 @@ class Request implements ArrayAccess
      *   this method will return true if the request matches any type.
      * @return bool Whether or not the request is the type you are checking.
      */
-    public function is($type)
+    public function is($type, $args = [])
     {
         if (is_array($type)) {
             $result = array_map([$this, 'is'], $type);
@@ -628,7 +628,7 @@ class Request implements ArrayAccess
         }
 
         if (!isset($this->_detectorCache[$type])) {
-            $this->_detectorCache[$type] = $this->_is($type);
+            $this->_detectorCache[$type] = $this->_is($type, $args);
         }
 
         return $this->_detectorCache[$type];
@@ -651,11 +651,12 @@ class Request implements ArrayAccess
      *   this method will return true if the request matches any type.
      * @return bool Whether or not the request is the type you are checking.
      */
-    protected function _is($type)
+    protected function _is($type, $args)
     {
         $detect = static::$_detectors[$type];
         if (is_callable($detect)) {
-            return call_user_func($detect, $this);
+            array_unshift($args, $this);
+            return call_user_func_array($detect, $args);
         }
         if (isset($detect['env']) && $this->_environmentDetector($detect)) {
             return true;
```

**Comment by jadb** [comment]
I submitted a PR #8550 (first was opened and/or named against the wrong branch) that if merged, would allow the detectors to be added in `bootstrap.php` for example.

@mirko-pagliai I like the idea a lot, and definitely see myself using it! thanks for proposing it :+1:

**Comment by mirko-pagliai** [comment]
very nice work @jadb thanks to you.

**Comment by mirko-pagliai** [comment]
Just one more thing and then we are.
I had also considered a method like `isHere()` (see code), because I imagined something like this:

    $this->request->isHere(['_name' => 'homepage_route']);

Maybe is it useless? What do you think about?

**Comment by jadb** [comment]
@mirko-pagliai, while helper methods are nice to have, some are more useful than other. Maybe if you put up a clear list of the ones you have in mind by opening a new PR against the app skeleton (that's where I believe the custom detectors will go), then we can all chime in and make a decision on which should be part of the officially distributed code.

**Comment by mirko-pagliai** [comment]
Thanks. So my PR can be closed :-)

**Pull request closed without merging**